### DISCUSSION TITLE
getting_started/first_3d_game/03.player_movement_code

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#78] [User-contributed notes]
# getting_started/first_3d_game/03.player_movement_code

It's time to code! We're going to use the input actions we created in the last part to move the character. Right-click the Player node and select Attach Script to add a new script to it. In the pop...

https://docs.godotengine.org/en/stable/getting_started/first_3d_game/03.player_movement_code.html

**Comment by orchard800** [comment]
Shouldn't the CameraPivot be attached to the player somehow?

**Reply by Calinou** [reply]
In this demo, the camera has a fixed position and covers the entire level (similar to minigames like Mario Party). Therefore, it doesn't need to follow the player.

**Comment by Elliott-Tanner** [comment]
When I follow these instructions, I end up with the camera looking directly down, with no angle. How should I fix this?

**Reply by discotwig** [reply]
"Select the CameraPivot and rotate it -45 degrees around the X axis (using the red circle). You'll see the camera move as if it was attached to a crane."
You can press E to enter Rotate Mode and use the red arc to rotate downward.

**Reply by Lensco825** [reply]
Had the same problem. I set the camera node's y-axis position to 18m, then you set it's projection to "Orthogonal" with it's size set to 19m. 
![image](https://github.com/user-attachments/assets/c4dc214a-d3a1-448f-8efb-8e9973f3207c)

**Reply by Deadly-Cicada** [reply]
This happened to me as well. I found that I was rotating the "Camera" and not the "CameraPivot".

When we moved the "Camera" back 19 units we should have left the "CameraPivot" in the center of the floor where the character was at. If we then move the red arc on "CameraPivot" our "Camera" will use the origin point of the pivot to move as opposed to its new location at -19Y.

**Reply by FrackinFamous** [reply]
Both comments have the same issue. You are moving the camera in the negative Y.  You should be moving it back, if you use the red arrow and pull it back as the docs suggest, you are moving along the Z axis in the positive direction. So a -19y that should be a +19z is the error not the docs. After sliding the camera back you rotate the thusly named CameraPivot which pivots the camera that is behind the player to a 45degree angle up and behind the player as if it were on a production boom in the movies. Camera3D transform: position 0x, 0y, 19z. CamerPivot transform: -45x, 0y, 0z.

No other transforms or settings need to be changed to set the camera this way.

**Comment by Wadahell** [comment]
My character doesn't move, it only turns to the side of the respective input 
Tried moving the floor, the char, the collision sphere, didn't work

Possible relevant info: Camera dimentions are a little different from the article..
![image](https://github.com/user-attachments/assets/841a6378-8bc3-45da-8572-d86d291746a6)

**Reply by mtt-sp** [reply]
Had the same issue, have you checked your code? The turning should mean the inputs are properly setup, check the grounded velocity and or the code prior to move_and_slide. Maybe the var's are not correspondingly typed or you wrote: .y instead of .x or .z like i did.

**Reply by Wadahell** [reply]
Yes, like 3 times, at this point i'm believing starting again from scratch is the best option
Did your's eventually worked?

**Reply by mtt-sp** [reply]
yeah

**Reply by Wadahell** [reply]
Found it! it was some indentation on the code, i gave up finding it and used the code at the end of the page
Also, theres no code for jump

func _physics_process(delta):
...
	if Input.is_action_pressed("jump"):
		direction.y += 1
		
used this and only partially worked (it did something

**Reply by kynoch** [reply]
This happened to me too. I diffed my code with the example and it matched exactly, but after printing the target velocity I noticed the y value was increasing steadily.

I think what happened was the player model was inside the floor. Raising it on the y axis in the 3D mode fixed it at any rate.

**Reply by GlitchyPixel** [reply]
Yeah I faced it too, found it was just a problem in the placement on the "Player" object. Increase it's y transform to be above the ground when starting the game and it get's fixed. The problem was because the Player object was actually kind of being "embedded" onto the ground.

**Reply by ganavasco** [reply]
I had a similar problem because I used Input.is_action_just_pressed instead of Input.is_action_pressed, so my character would move an inch and turn to each direction, but wouldn't keep moving. Took me a bit to find haha

**Comment by BryceDMonaco** [comment]
> Here, we're going to make all calculations using the _physics_process() virtual function. Like _process(), it allows you to update the node every frame

Isn't this incorrect? My understanding is _physics_process() runs at a fixed rate independent of visual frame rate. Maybe explaining fixed rate updates and physics frames is out of the scope of this tutorial, but I think it can lead to confusion to those who are completely new to game dev.

**Comment by bruno-normande** [comment]
```gdscript
# The downward acceleration when in the air, in meters per second squared.
@export var fall_acceleration = 75
```

75 is much greater than earth's gravity (~9.8m/s2). Is that so that the player falls faster? Or is there some mechanical reason for that? (or is the comment wrong?)

**Reply by ghost** [reply]
Video games in general and historically, not just in Godot, have very high gravity. You can see this in every game from original super mario (~22.1) or minecraft (~32.7). 75 is rather high but there is also no acceleration here so that's something to consider, as well as what jumping actually does in the game, that is killing creeps, and for that it needs to be very high and responsive in order to land quickly after jumping.

**Comment by vmorris** [comment]
There's no instruction to raise the Y position of the camera.

**Reply by ragn4rok234** [reply]
Moving the Y position of the camera isn't required. Changing the X rotation on the CameraPivot node to -45 raises the camera to the correct position

**Reply by vmorris** [reply]
This was not my experience. The camera node is at 0,0,0, rotating it -45 X doesn't raise it up at all.

**Reply by ragn4rok234** [reply]
You're not supposed to rotate the camera node itself, you need to rotate the CameraPivot Node that the camera is a child of.
Also the camera node itself should be at around (0,0,19) not (0,0,0) based on the instructions in the lesson.

**Reply by vmorris** [reply]
Yes, you're right about the position around (0,0.19). I will try again with the CameraPivot Node instead. Thanks!

**Reply by guiYudi** [reply]
I had to move it up a bit. My CameraPivot (marker3D) is at (0,15,19) rotated to (-45º, 0, 0) and my Camera3D is at (0,0,0) (with no rotation). If I position the CameraPivot (marker3D) at (0,0,19) it only renders the top half of the viewport, and when the player moves down, it clips into an invisible area (probably clipping into the camera).

**Reply by FrackinFamous** [reply]
I think the boom arm analogy is escaping people by the time they get to positioning the nodes. The reason we make the Camera3D a child of CameraPivot is this. The marker will remain with the subject of the camera.. the player. The marker DOES NOT have it's position changed. Now we extend the boom. We slide the Camera3D back (positive Z direction 19z). If we now rotate the CameraPivot it moves the boom into the air on a 45degree angle. 

CameraPivot position 0,0,0 rotation -45, 0, 0
    Camera3D(child of CameraPivot) position 0,0,19 rotation 0,0,0.

**Comment by guiYudi** [comment]
```GDScript
if not is_on_floor(): # If in the air, fall towards the floor. Literally gravity
    target_velocity.y = target_velocity.y - (fall_acceleration * delta)
```

Why not just `target_velocity.y -= (fall_acceleration * delta)` ?

**Reply by HeadedBranch** [reply]
write a pull request if the code works and you think it is better coding practice

**Reply by BluesAdam** [reply]
@Headedbranch225 we are all here to inquire and learn. The worst thing you can do is sass someone for being curious and asking a question. Please consider responding to questions in a more constructive and encouraging manner.

I wonder the answer to this question as well. So if anyone could please chime in.

**Reply by ImminentTofu** [reply]
I think it's because if you miss the "if not is_on_floor():" your character constantly builds up speed downward and never resets. So when you go to jump it'll probably keep you glued to the ground depending on how you code it. Also might just be a best practice thing; if there's a cliff and you walk off it, you'll be smacked into the ground at mach 5 instead of at the fall_acceleration.

**Reply by abrossimow** [reply]
Practically it still is the same code. The original annotation `target_velocity.y = target_velocity.y - (fall_acceleration * delta)` was (probably) used here to point out the connection for all three axis:
- First we set the x-axis with target_velocity.x = ...
- Next we set the y-axis
- And finally we set the z-axis.

**Comment by ItsCorruptivity** [comment]
My forward backword movement is inverted

**Reply by ItsCorruptivity** [reply]
I checked out the camera and code everything was right

**Reply by guiYudi** [reply]
Did you map the keys correctly in the Input Mapping? `move_forward` should be mapped to up and `move_back` to down keys.

**Reply by BluesAdam** [reply]
In Godot, forward is -1 and backwards is 1 when it comes to y-axis. You might want to double check. That caught me by surprise as well.

**Reply by BluesAdam** [reply]
Sorry, meant to say z-axis.

**Comment by BrunoProsdocimo** [comment]
My code doesn't do anything. I'm trying to code with C# using Visual Studio Code and the code seems to be connected to the project correctly as when I click the "script" icon on the Player scene it opens VS Code with the all the code in it. When I play the game however the code seems like it doesn't do anything. I've tried debugging it by printing lines in output when the buttons in the input are pressed but that also doesn't happen. I've also added the proper inputs in the editor. I've also tried just copying the entire code from this page into VS Code and that didn't work. 

At this point I'm assuming it's something with the editor, and I've even downloaded the Godot Engine .NET from their website, reloaded the project and it also didn't work. I don't know what's going on, can anybody give me a hand?

**Reply by Menekrr** [reply]
Hey!

If I'm understanding you correctly, you were saying that when you ran the game (after choosing to run the Main scene), you can't move the character, and any print() statements you put into the ```_physics_process()``` function also don't print anything.

If that's the case, I encountered the exact same issue, but not sure if the circumstances were the same as yours.

When I instantiated a Player scene into the Main scene, I accidentally chose the ```art/player.glb``` instead of the ```player.tscn```. I noticed something was weird when I saw that my Player scene was capitalized (capital 'P'), but the instantiated player in ```Main``` was *not* capitalized (lower-case 'p').

When I deleted the "player" instance and instead instantiated a "Player", my problems went away.

Hope that might help you! Could you try it out?

**Reply by SarahRamage** [reply]
Thank you so much that's exactly what I did also

**Reply by rdepew** [reply]
@Menekrr, this error bit me also: I chose the `player.glb` file instead of the `player.tscn` file. Deleting the small-p instance and instantiating a new Player using the `player.tscn` file fixed it for me.

**Reply by Nedus** [reply]
@Menekrr exactly my problem too! Weird that by wrongly instantiating the script it would also visualize the mesh, this was what threw me off.

**Reply by nklisch** [reply]
This is what I did as well. I was very confused for longer then I'd like to say....lol

**Reply by eatthischair** [reply]
THANK YOU oml I had no idea what was going on
For future reference to anyone who dragged player.glb over the Player instance, you don't have to delete the instance, just drag player.gd over it and everything works

**Comment by GistOfSpirit** [comment]
> The `CharacterBody3D.is_on_floor()` function returns `true` if the body collided with the floor in this frame. That's why we apply gravity to the `Player` only while it is in the air.

What counts as "floor" though? I think it would be useful to have a short explanation of that.

As far as I can tell, the big rectangle we created doesn't have any special properties that would label it as "floor" (as opposed to "wall", "character" etc.). I presume that any collision shape is a floor as far as `is_on_floor` is concerned?

**Reply by Calinou** [reply]
> I presume that any collision shape is a floor as far as is_on_floor is concerned?

Whether the CharacterBody is considered to be "on floor" is determined by the collision normal, not the collision shape type or its metadata.

The maximum collision normal value that is considered "on floor" can be configured by adjusting **Floor > Max Angle** in the inspector. This is the steepest angle that characters will be able to walk on without being considered airborne (and therefore sliding off the ramp).

**Reply by FrackinFamous** [reply]
Hover the function or ctrl/cmd + click it. The hover gives the short explanation you crave here and the click gives you the docs entry in editor with links. That is why everything is not explained in great detail. This is not meant to come across rude or "JUST READ THE DOCS" sounding. It's literally just the first line of help for one not used to coding. Always check tooltips first then the docs for more in depth explanation.

**Comment by RobbiewOnline** [comment]
I have this mostly working, the player is visible on the floor, stationery.  When I click any of the movement keys the character moves in that direction, then falls through the floor.

I thought it must be the `CollisionShape3D` node being lower than the player, so I moved it up.  It made no difference, so I moved the floor even further down.

When I launch the game I now see the player fall and stop at the floor.  But when I press a key to move, the player still falls through the ground again.

‼️FYI ‼️

A number of times I went to run the game and I'd see a grey window, not my game .. took me a while to notice that it was when I was tinkering with the Player, because the 'main' scene tab isn't selected (player tab was), so the scene being 'run' was not the 'main' but the 'player' scene which has no camera🤦‍♂️

**Reply by RobbiewOnline** [reply]
And to be clear I did copy + paste the code above, which has the block

```
	if not is_on_floor():
		target_velocity.y = target_velocity.y - (fall_acceleration * delta)

	velocity = target_velocity
	move_and_slide()
```

**Reply by RobbiewOnline** [reply]
Okay, clear head and found my issue - I'd set the size of my MeshInstance3D correctly, but missed the steps to set my `CollisionShape3D`, which now makes sense...

When the game first started the character would fall precisely onto the default CollisionShape3D, as soon as I moved the character would fall off the `CollisionShape3D` area.

For others, this is the step I missed for `CollisionShape3D`...

> A box's wireframe appears in the viewport with three orange dots. You can click and drag these to edit the shape's extents interactively. We can also precisely set the size in the inspector. Click on the BoxShape3D to expand the resource. Set its Size to 60 on the X-axis, 2 for the Y-axis, and 60 for the Z-axis.

**Comment by Userunmanned** [comment]
The movement works just fine with GDScript, but it doesn't work with C#. I am using the exact code provided.

**Reply by Userunmanned** [reply]
I have .NET 9.0 on my computer if that is needed

**Reply by Calinou** [reply]
Please ask support questions on other community channels: https://godotengine.org/community/

**Reply by Userunmanned** [reply]
But, shouldn't it work nonetheless?

**Comment by balloonbrain** [comment]
why use .ZERO instead of initialising the Vector3D like: "Vector3D(0.0, 0.0, 0.0)" is there a reason or is it just clarity and ease of use?

**Reply by Calinou** [reply]
It's just for clarity. You can also use `Vector3()` which is the same as `Vector3.ZERO`.

**Reply by balloonbrain** [reply]
thanks!

**Comment by sryimnew2this** [comment]
These are hilariously easy to follow; thanks, I love it!
One question, though, please? 
How do I do this with C++ using gd extensions? I've been shamelessly struggling for hours trying to figure it out

(I have a bouncy sprite moving across the screen in 2d thanks to bouncysprite.gdextension, & I tried to modify it to make a movecharacter.gdextension for 3d using vector3 & character_body3d, but i'm honestly lost)

I am grateful for any and all assistance! Thanks

**Reply by sryimnew2this** [reply]
I'm using visual studio 2019 if it helps, at all

**Reply by Userunmanned** [reply]
I literally directly followed the tutorial and copy and pasted the C# code and it still didn't work.

Your mocking is kind of funny though, I have to admit.

**Reply by sryimnew2this** [reply]
no no, I was being genuine; I was worried before I started that it would be difficult to follow (I tried doing the GDExtensions last year and somewhat struggled) but this in comparison is 100x easier and quite enjoyable!

Would be great if the c++ gdextension was a bit easier to use, though. (Although I am a fresh novice)

**Comment by playhorsejr** [comment]
Im getting an error code 
player.gd:25 @ _physics_process(): Node not found: "Pivot" (relative to "/root/main/player").
what can i do to fix this, using gdscript and copy pasted the code into my script and still didnt work

**Reply by ryanttb** [reply]
1. Did you include the dollar sign? `$Pivot.basis`
1. Is the Pivot node directly under your Player node?

Please post your script.

**Reply by playhorsejr** [reply]
```gdscript
extends CharacterBody3D

# How fast the player moves in meters per second.
@export var speed = 14
# The downward acceleration when in the air, in meters per second squared.
@export var fall_acceleration = 75

var target_velocity = Vector3()


func _physics_process(delta):
	var direction = Vector3()

	if Input.is_action_pressed("move right"):
		direction.x += 1
	if Input.is_action_pressed("move left"):
		direction.x -= 1
	if Input.is_action_pressed("move back"):
		direction.z += 1
	if Input.is_action_pressed("move forward"):
		direction.z -= 1

	if direction != Vector3():
		direction = direction.normalized()
		$Pivot.basis = Basis.looking_at(direction)

	# Ground Velocity
	target_velocity.x = direction.x * speed
	target_velocity.z = direction.z * speed

	# Vertical Velocity
	if not is_on_floor(): # If in the air, fall towards the floor. Literally gravity
		target_velocity.y = target_velocity.y - (fall_acceleration * delta)

	# Moving the Character
	velocity = target_velocity
	move_and_slide()
```

**Comment by ryanttb** [comment]
One comment, in the Previous lesson, I moved my Player's CollisionShape3D Position up too high, to y = 1m. It kept falling halfway through the floor. I changed it to 0.5m and it's fine now.

**Reply by n-hdev** [reply]
Yes, I had this too, so I went back to look at the collision shape and I think I'd been using the axis markers rather than the shape outline to position the shape in the earlier tutorial. 

On the other hand, half through the floor means you just see its eye, very creepy, much scarier... almost got a newspaper out to squish it dead on the screen.

**Comment by molorosh** [comment]
So the player didn't move for me at first.  But then I noticed that in the Node tree view my player was not showing a script / document icon - and then I realised - I had added the player.glb model file instead of the _scene_ file.  I removed the node and added the child scene instance of scene.tscn and then it all worked great.

I hope that helps other people. :-)

**Reply by slegetank** [reply]
This is really helpful! Thank you for saving me time!

**Reply by SaleemVFX** [reply]
!!!!! THISSSS I've been frustrated why it wasn't working lol . Thanks much

**Comment by burckd** [comment]
I could not found any CameraPivot node to add while using Godot Engine v4.4
Instead I added a Node3d and named it Pivot, so far everything works fine 

I might be missing something, but I hope this helps.

**Reply by Calinou** [reply]
There is no CameraPivot node in Godot. This page tells you to create a Marker3D node and rename it CameraPivot instead:

> Create a new [Marker3D](https://docs.godotengine.org/en/stable/classes/class_marker3d.html#class-marker3d), and name it `CameraPivot`. Select `CameraPivot` and add a child node [Camera3D](https://docs.godotengine.org/en/stable/classes/class_camera3d.html#class-camera3d) to it.

**Reply by burckd** [reply]
Thanks for your time, sorry. I totally missed something. Should I just delete the comment?

**Reply by Calinou** [reply]
> Should I just delete the comment?

You can leave it here if you want, or remove it depending on what you're most comfortable with.

**Comment by gunererd** [comment]
Is there a particular reason or advantage to put Camera3D under a pivot node? Why don't we position and rotate camera directly?

**Reply by KonscienceGit** [reply]
The Pivot node allow to rotate the translated Camera, in a way that the camera always look at the Pivot position no mater the rotation.
In this example, the Camera node is moved backward by 19 meters, so when we later apply the -45° rotation to the parent Pivot node, the Camera doesn't just rotate on itself, but instead rotate around it's origin (which is 19 meters in front of it).

In short, applying the rotation on the pivot allow to handle the camera like if it was on a selfie stick.

**Comment by TheDeadMeasure** [comment]
When I try to pull the Camera3D on the z axis (with it selected in the Scene panel), the view shown by the camera barely leaves the character's shape. How do I fix that? (To get the effect similar to that shown in the tutorial pics, I have to pull the camera about 850 units away.)
![Screenshot 2025-05-05 154756](https://github.com/user-attachments/assets/c490b012-17d2-4210-b244-f353d87090eb)

**Reply by KonscienceGit** [reply]
It seems your Camera FOV (field of view) setting in the right panel (inspector) was changed to 1°, which is probably why it feels like overly zoomed onto the character. You can press the circling arrow symbol to revert it to the default value (75°).

**Reply by TheDeadMeasure** [reply]
Thank you much!

**Comment by Gardioz** [comment]
Hi!

I've been stuck at running the scene but only getting a grey background. Think it's about something regarding directional light or the player.glb file. Tried many configs in both but can't get it working properly, can anybody help?

![image](https://github.com/user-attachments/assets/43e354d4-f607-415d-8ad4-7bbe7abd4dee)
![image](https://github.com/user-attachments/assets/65a00e69-692b-41f8-83b0-830a69dacc36)

**Reply by Calinou** [reply]
You can move around in the running project by clicking the **3D** button at the top of the game window, then clicking the camera icon in the toolbar on the right (still on the game window). After doing that, you can hold the right mouse button and use WASD to move around while the game is running. This is useful to diagnose issues like that, so you can see what went wrong exactly.

**Comment by KermitThePog** [comment]
Is there a reason to multiply fall acceleration by delta when using ``` _physics_process() ```, given that uses fixed time intervals?

**Reply by Calinou** [reply]
You can change the frequency at which `_physics_process()` runs in the Project Settings (**Physics Ticks per Second**). Using `delta` ensures that if you change this value later on, it won't make gravity stronger or weaker in the project. In general, it's also good programming practice to use delta time and avoid hardcoding around a 60 Hz physics assumption.

**Comment by Azjur** [comment]
Hello guys, I have been learning Godot a week now, i have finished the 2D game and now trying to do the 3D game as well. I am almost at the end, but i suddenly notice that my character is not faceing the direction it moves.
I guess that this part of the code should handle it:

$Pivot.basis = Basis.looking_at(direction)

But for me it's mot working? Or i am doing something wrong?

Please help and thank you.

**Reply by Azjur** [reply]
Haha....no help needed, i actually figured it out. My character model was not a child of pivot node. Went back to the top and trying to check the hierarchy.
Anyway it's great to trying to troubleshoot and then actually find the issue yourself. That's the learning curve right there.

**Comment by chihwahli** [comment]
I am trying to understand the instructions from on "In the top view, make sure your Camera3D is selected and move the camera about 19 units on the Z axis (drag the blue arrow)." But I really have no clue what is meant, what am I supposed to do? It all looks good to me by default. Why should I change it? Instructions are not clear.

**Reply by chihwahli** [reply]
Ah, got it, text did not ring a bell.

**Comment by chihwahli** [comment]
$Pivot.basis = Basis.looking_at(direction) gives the error:

invalid assignment of property or key 'basis' with value of type 'Basis' on a base object of type 'null instance'

It is typed correctly, I even copied and pasted it the 2nd time.

**Reply by ronnierap** [reply]
The hint: "........of type 'null instance" tells us you probably do not have a"Pivot" under the Player node?
![Player Scene](https://docs.godotengine.org/en/stable/_images/toggling_visibility.webp)

If one would have seen: 
Invalid assignment of property or key 'basis' with value of type 'Basis' on a base object of type 'Node'.

The type of the Pivot is just Node and not Node3D

But yours is probably the Pivot node not present under Player (see if the name matches "Pivot")

**Comment by juanjvmedrano** [comment]
hi guys, by any chance any of you had a problem were it does a little zoom once you click any of the inputs at the beginning, move_left, etc. I didn't touch any of the camera's properties

**Comment by ShadowAngel27** [comment]
My player won`t move at all. Even after copying full code higher. Tried lifting player so collision box isn`t in floor
Thank you!

**Reply by ShadowAngel27** [reply]
#Find the Solution

**Comment by owenn** [comment]
My player was not moving, added breakpoints in the _physics_process. It was not been called. I realised I had selected the gdb file  instead of the player scene when I instantiated it.

**Comment by v-kydela** [comment]
Is the player supposed to be half sunken into the floor?

<img width="566" height="280" alt="image" src="https://github.com/user-attachments/assets/c57059e4-6b16-4852-a3ab-8ae5a853cafb" />

**Reply by SriramAnanth2912** [reply]
yeah, I have the same scenario. maybe it should be as such.
if you move the ground y to -2m you can see the shadow of it but when you are squashing the creeps we have to land and no shadow to be there then we get as given here.

**Comment by mjrearden02** [comment]
In the view port my player model is on the ground but when I run the game it is floating.

**Reply by mjrearden02** [reply]
When I moved the collision sphere above my model in the player scene the model was one the ground fine but the hitbox was too high now.

**Reply by mjrearden02** [reply]
nvm guy in the prev section had the fix. In the main scene I only moved the mesh floor down so the collisionnode was at the same height.

**Comment by shangyuq** [comment]
Within *_physics_process()* method, I want to explictly show the instance of *is_on_floor()* method by *$Player.is_on_floor()*.
But Debugger told me that, *Attempt to call function 'is_on_floor' in base 'null instance' on a null instance.*
Why "$Player" is not instantiated? Why is okey to use *is_on_floor()* directly without prefix in example?
Thank you!

**Reply by BakaBananana** [reply]
Late reply, so for anyone else who is also confused, $Player looks for the child node with the name "Player" .  How we refer any node inside $ is similar to how we navigate in terminals.  Here $"." would be the correct reference to the player node but it's redundant to use.

**Comment by barry117** [comment]
Hello all just a quick one i have gone through the lessons up until the camera placement all went smooth except my character movement kind of, the only problem is that to move i must continuously tap the direction of travel to move, i am unable to press and hold to move. any ideas?

**Reply by Calinou** [reply]
Make sure your player movement code is in `_physics_process()`, not `_input()`.

**Reply by barry117** [reply]
Thank you!! you are a legend while this wasn't the issue it pushed me to the right place i had is_action_just_pressed instead of is_action_pressed

**Comment by RhinoDany** [comment]
I am getting this error and my code is exactly as the example. Im using GDScript.

"Error setting property 'velocity' with value of type Vector3."

Which does bother me, in the code we never declare the velocity variable until the end without a type.

I tried adding the following before that line, but when running it the player would only change directions but not move.

	# Moving the Character
    var velocity = Vector3.Zero
	velocity = target_velocity
	move_and_slide()

**Reply by RhinoDany** [reply]
<img width="1348" height="606" alt="image" src="https://github.com/user-attachments/assets/bf8178f0-7ac0-454f-9b56-4c278e183725" />

**Reply by RhinoDany** [reply]
Nevermind Im an idiot I nested my player node in Character 2D instead of a 3D. Re-do!

**Comment by admin** [comment]
*Originally posted by [**kristianpayne1**](https://github.com/kristianpayne1) on 2026-06-04 20:25:23 UTC*

> Be careful to use `is_action_pressed()` (instead of `is_action_just_pressed()` which the editor suggests first). That tripped me up for a while!

**Reply by admin** [reply]
*Originally posted by [**zelaan**](https://github.com/zelaan) on 2026-06-18 00:50:54 UTC*

> both useful and both have their useings in specific functions.

**Comment by admin** [comment]
*Comment originally posted by [SamuelThomasLambert](https://github.com/SamuelThomasLambert):*

> my Player kept falling through the floor as soon as they moved from the starting position. I had the size of the MeshInstance3D set correctly but some reason it didn't work properly, deleting the mesh and creating it again solved the problem.

**Reply by admin** [reply]
*Comment originally posted by [Lokenoth](https://github.com/Lokenoth):*

> I had this exact same issue and I believe its an issue with hierarchies. you probably like me had the original Mesh under the Collision without noticing it. Swapping the two round fixed mine. 
>
> So
>
> Ground
>          MeshInstance3D
>          CollisionShape3D
>
> Should be the correct order.

**Reply by admin** [reply]
*Comment originally posted by [SamuelThomasLambert](https://github.com/SamuelThomasLambert):*

> Thats quite likely! I'm glad you worked out what it was

**Comment by gitoli56** [comment]
Why using a separate Pivot object to rotate the camera knowing the Camera3D also has a transform property and can be rotated using it?

**Comment by rdepew** [comment]
Think of the Pivot as a selfie stick.

________________________________

**Comment by sen507451-sys** [comment]
extends CharacterBody3D

# How fast the player moves in meters per second.
@export var speed = 14
# The downward acceleration when in the air, in meters per second squared.
@export var fall_acceleration = 75

var target_velocity = Vector3.ZERO


func _physics_process(delta):
	var direction = Vector3.ZERO

	if Input.is_action_pressed("move_right"):
		direction.x += 1
	if Input.is_action_pressed("move_left"):
		direction.x -= 1
	if Input.is_action_pressed("move_back"):
		direction.z += 1
	if Input.is_action_pressed("move_forward"):
		direction.z -= 1

	if direction != Vector3.ZERO:
		direction = direction.normalized()
		# Setting the basis property will affect the rotation of the node.
		$Pivot.basis = Basis.looking_at(direction)

	# Ground Velocity
	target_velocity.x = direction.x * speed
	target_velocity.z = direction.z * speed

	# Vertical Velocity
	if not is_on_floor(): # If in the air, fall towards the floor. Literally gravity
		target_velocity.y = target_velocity.y - (fall_acceleration * delta)

	# Moving the Character
	velocity = target_velocity
	move_and_slide()
